Persistent data structures by im-rs

Eh2406 opened the pull request:
There has been a long standing "TODO: We should switch to persistent hash maps if we can" in the resolver. This PR introduces a dependency on [im-rs](https://github.com/bodil/im-rs/issues/26) to provide persistent hash maps. It then uses `im_rc::OrdSet` to store the priority list of dependencies, instead of `std::BinaryHeap`, as cloning the `Heap` was one of the most expensive things we did. In Big O terms these changes are very big wins, in practice the improvement is small. This is do to a number of factors like, `std::collections` are very fast, N is usually only in the hundreds, we only clone when the borrow checker insists on it, and we use `RC` and other tricks to avoid cloning.

I would advocate that we accept this PR, as:
- Things are only going to get more complicated in the future. It would be nice to have Big O on our side.
- When developing a new feature finding each place to add `RC` should be an afterthought not absolutely required before merge. (cc #6129 witch is stuck on this kind of per tick performance problem as well as other things).
- As the code gets more complex, making sure we never break any of the tricks becomes harder. It would be easier to work on this if such mistakes are marginal instead of show stoppers. (cc #5130 a bug report of one of these bing removed and affecting `./x.py build` enough to trigger an investigation).
- The resolver is already very complicated with a lot of moving parts to keep in your head at the same time, this will only get worse as we add  features. There are a long list of tricks that are *critical* before and `~5%`  after, it would be nice to consider if each one is worth the code complexity. (I can list some of the ones I have tried to remove but are still small wins, if that would help the discussion).

Overall, I think it is worth doing now, but can see that if is not a slam dunk.

dwijnand commented on src/cargo/core/resolver/types.rs:
```diff
@@ -299,48 +305,45 @@ impl Ord for DepsFrame {
 /// use (those with more candidates).
 #[derive(Clone)]
-pub struct RemainingDeps(BinaryHeap<DepsFrame>);
+pub struct RemainingDeps(usize, im_rc::OrdSet<(DepsFrame, usize)>);
```
I've need trying to find good names for the two `usize` fields, and failing.  Can you think of good names and label the fields in code?  It think it would help understanding the code.

Eh2406 commented:
I added some comments, and some variable names. Does that help?

dwijnand commented:
Yeah, it does, thank you!

I had assumed that the increment was relevant to the ordering.  If it's just a question of not making elements disappear, why not just have a wrapper with a custom `Eq` that's never `true`?  Say we call it `NeverEqDepsFrame` that would re-simplify `RemainingDeps`  to `RemainingDeps(im_rc::OrdSet<NeverEqDepsFrame>)`.

dwijnand commented:
Outside of that, I'm convinced by your arguments so I'm happy to land this. 😄

Eh2406 commented:
... let me try that. BRB.

Eh2406 commented:
That is what I remembered, lots of the test fail to pass. According to [Ord "Implementations of PartialEq, PartialOrd, and Ord must agree with each other."](https://doc.rust-lang.org/std/cmp/trait.Ord.html) So I changed `PartialEq` to `false` and added `.then_with(|| Ordering::Less)` to `Ord`. I don't know why that did not work, but it dose not. Not with `im_rc::OrdSet` nor with `std::BTreeSet`. I think it may brake the antisymmetry definition of [PartialOrd](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html).

dwijnand commented:
Yeah `im_rc::OrdSet` requires a total `Ord`, so use my idea.  I agree with you it's probably breaking the antisymmetry law that's breaking it.

Eh2406 commented:
It also needs to be deterministic, so we can't use the memory address to duplicate.

alexcrichton commented:
This looks pretty awesome to me, having drop-in replacements is indeed really nice!

Out of curiosity, have you taken a look at the perf as-is today? (just to make sure we're not accidentally having a regression)

Eh2406 commented:
According to the discussion in https://github.com/bodil/im-rs/issues/26 I have had a version of this branch since Jun 21. I have at various times done detailed analysis of [micro benchmark performance](https://github.com/bodil/im-rs/issues/26#issuecomment-400025049) and [experiments with cargo performance](https://github.com/bodil/im-rs/issues/26#issuecomment-399240858). Overall both `std` and `im-rs` are very fast. Definitively saying witch is the better for a given situation is *hard* and very dependent on exactly what you are bench-marking. There is a large range where there performance are functionally identical. This PR only includes the changes where it was a measurable improvement, when I last seriously tested on the benchmarks I happened to be using. I seriously doubt there is a big change in perf as-is today

However I have not rerun the perf recently, if you have something you would like me to try let me know. (Servo is ridiculously slow on my computer, so not that.)

alexcrichton commented:
@bors: r+

Ok that sounds reasonable to me! I definitely looks like you've been far more thorough here than I would have been :)

bors commented:
:pushpin: Commit 1699a5b1ce197596dae286fd39d48e7b91b3ed5f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 1699a5b1ce197596dae286fd39d48e7b91b3ed5f -->

bors commented:
:hourglass: Testing commit 1699a5b1ce197596dae286fd39d48e7b91b3ed5f with merge c93e84716d279cdff11304cef627f8eb72dd7e9a...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/builds/20469183), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/458074374?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing c93e84716d279cdff11304cef627f8eb72dd7e9a to master...
